test/org/javarosa/xpath/expr/XPathPathExprCurrentTest.java

```diff
@@ -0,0 +1,100 @@
+    private FormDef formDef;
+    private FormEntryController formEntryController;
+
+    @Before
+    public void setUp() {
+        FormParseInit fpi = new FormParseInit(r("relative-current-ref.xml"));
+        formDef = fpi.getFormDef();
+        formDef.initialize(true, new InstanceInitializationFactory());
+        FormEntryModel formEntryModel = new FormEntryModel(formDef);
+        formEntryController = new FormEntryController(formEntryModel);
+    }
+
+    /**
+     * current() in a calculate should refer to the node it is in (in this case, /data/my_group/name_relative).
+     * This means that to refer to a sibling node, the path should be current()/../<name of sibling node>. This is
+     * verified by changing the value of the node that the calculate is supposed to refer to
+     * (/data/my_group/name) and seeing that the dependent calculate is updated accordingly.
+     */
+    @Test
+    public void current_in_calculate_should_refer_to_node() {
+        formEntryController.jumpToFirstQuestionWithName("name");
+        StringData nameValue = new StringData("Bob");
+        formEntryController.answerQuestion(nameValue, true);
+
+        StringData relativeName = (StringData) formDef.getFirstDescendantWithName("name").getValue();
+        assertEquals(nameValue.getValue(), relativeName.getValue());
+    }
+
+    /**
+     * current() in a choice filter should refer to the select node the choice filter is called from, NOT the expression
+     * it is in. See https://developer.mozilla.org/en-US/docs/Web/XPath/Functions/current -- this is the difference
+     * between current() and .
+     *
+     * The behavior of current() in a choice filter is verified by selecting a value for a first, static select and then
+     * using that value to filter a second, dynamic select.
+     */
+    @Test
+    public void current_in_choice_filter_should_refer_to_node() {
+        formEntryController.jumpToFirstQuestionWithName("fruit");
+        Selection fruitSelection = new Selection(1);
+        fruitSelection.attachChoice(formEntryController.getModel().getQuestionPrompt().getQuestion());
+        formEntryController.answerQuestion(new SelectOneData(fruitSelection), true);
+
+        SelectOneData fruitSelectionValue = (SelectOneData) formDef.getFirstDescendantWithName("fruit").getValue();
+        assertEquals("blueberry", fruitSelectionValue.getDisplayText());
+
+        // === Variety question ===
+        formEntryController.stepToNextEvent();
+        // Collect calls getAnswerValue to populate dynamic choices
+        FormEntryPrompt varietyPrompt = formEntryController.getModel().getQuestionPrompt();
+        varietyPrompt.getAnswerValue();
+        ItemsetBinding dynamicChoices = varietyPrompt.getQuestion().getDynamicChoices();
+
+        // There are three blueberry varieties defined by the form
+        assertEquals(dynamicChoices.getChoices().size(), 3);
+
+        SelectChoice varietyChoice = dynamicChoices.getChoices().get(1);
```

Comment:
Are the next assertions really necessary to verify the behavior of `current()` in this context?

Reply:
I suppose not. This block is mimicking what a client does when actually making a selection. It feels more complete to be able to say with certainty that a choice can not only be saved but also successfully retrieved from the model after that but I suppose we could do that in a separate test.

Reply:
Sorry to be picky but it's a huge load of complex stuff that's going on. Anything we can remove will make the whole easier to understand for everyone.